File: 백준/Gold/3079. 입국심사/입국심사.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
)

var (
	reader = bufio.NewReader(os.Stdin)
	writer = bufio.NewWriter(os.Stdout)
)

func main() {
	defer writer.Flush()
	var left, right, n, m, res, person, mid int
	left = math.MaxInt
	fmt.Fscanf(reader, "%d %d\n", &n, &m)
	var times = make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Fscanf(reader, "%d\n", &times[i])
		if left > times[i] {
			left = times[i]
		}
		if right < times[i] {
			right = times[i]
		}
	}
	right = right * m
	res = right
	for left <= right {
		person = 0
		mid = (left + right) / 2
		for i := 0; i < n; i++ {
			person += mid / times[i]
		}
		if person >= m {
			right = mid - 1
			if res > mid {
				res = mid
			}
		} else {
			left = mid + 1
		}
	}
	fmt.Fprintf(writer, "%d\n", res)
}

File: 백준/Gold/3079. 입국심사/입국심사.py
import sys 
input = sys.stdin.readline 
sys.setrecursionlimit(10**6)
n, m = map(int, input().split())
times = []
for _ in range(n):
    times.append(int(input()))

left,right = min(times), max(times) * m #최소 : 한 번에 모조리 끝, 최대 : 모든 사람이 최장 시간 걸림
res = right 

while left <= right:
    person = 0 
    mid = (left + right) // 2
    for i in range(n):
        person += mid // times[i] #지금 구한 시간으로 통과 가능한 사람
    if person >= m:
        right = mid - 1
        res = min(res, mid)
    else:
        left = mid + 1
print(res)